### DISCUSSION TITLE
"IERC20" or "address" for "token" argument

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#147] [General]
The `create` functions expects a ERC-20 contract in the `token` argument. We're currently requiring users to import the `IERC20` defined in [prb-contracts](https://github.com/paulrberg/prb-contracts), but users might not like that. That is, they might not like the fact that in order to use Sablier's `v2-core`, they have to also import `prb-contracts`. The latter is effectively a peer dependency of the former.

See the discussion on [Twitter](https://twitter.com/PaulRBerg/status/1583789145205268481) for more thoughts and opinions.

**Comment by andreivladbrg** [comment]
Personally I prefer using `address`. This is also how I implemented the `token` argument in the [periphery](https://github.com/sablierhq/v2-periphery/blob/7a14e79ca53d6220f14324fa415f6046094b549e/src/AirStream.sol#L27).

I support the change.

**Reply by razgraf** [reply]
I think this was about *not* forcing the user depend upon an additional import from `prb-contracts`.

Especially for an interface which is usually already being used in most contracts as an import from e.g. openzeppelin.

For example [Solmate](https://github.com/transmissions11/solmate) doesn't even seem to include a bespoke IERC20 export.

**Reply by PaulRBerg** [reply]
That's right - the question is more so if we should switch to OZ rather than keep using (and I, maintaining) `prb-contracts`.

The only problem with importing and using OZ is that [they do not have custom errors][1]. If they only had them, it would be a no-brainer to start using OZ right away.

Re Solmate - I wouldn't use it as a guideline. It's a bad library from several reasons (I won't go into this here).

[1]: https://twitter.com/PaulRBerg/status/1586351027275632645

**Comment by PaulRBerg** [comment]
I'm leaning towards one of the following solutions:

1. Use `address` instead of `IERC20`.
2. Sick with `IERC20` but switch to OpenZeppelin.

I'm leaning towards the former, since OpenZeppelin is no longer a gold standard, and the type benefits aren't that great anyway. The function execution will just revert if an invalid token contract is provided - nothing will change in our own safety/ security assumption.

**Comment by PaulRBerg** [comment]
Locking, since @andreivladbrg ended up implementing the vanilla `address` type in PR https://github.com/sablierhq/v2-core/pull/176.

**Comment by PaulRBerg** [comment]
I'm unlocking this conversation because the recent proposal to re-export the PRBMath types (https://github.com/sablierhq/v2-core/issues/202) has gotten me thinking if it wouldn't be worth it to undo the changes that @andreivladbrg has done in https://github.com/sablierhq/v2-core/pull/176, and go back to using `IERC20` for the `token` argument?

A benefit of using `IERC20` would be that we would also be able to use it in the fees mapping, so that the keys are typed there, too.

We would simply re-export `IERC20` in a `Types.sol` file.

Obviously this is quite a bit different proposal to the PRBMath types, because most projects have an ERC-20 interface in their projects.

What do you think, @andreivladbrg?

**Comment by PaulRBerg** [comment]
I am reviving this discussion because I just stumbled upon a major argument in favor of using the `IERC20` interface file we import from `prb-contracts` as a type for the `token` fields in our streams and function arguments.

The argument is as follow - the stream is also an NFT, i.e. a "token", so not having a typed `token` field makes it easy to confound the ERC-20 "token" fields with the NFT token that the stream is. The issue is made worse by the fact that we are using an implementation of ERC-721 which refers to the ID of the NFT as `tokenId`.

What do you think, @andreivladbrg and @razgraf?

I also thought about whether it would be worth it to use OpenZeppelin's implementation - and my advise would be not to. My rationale is two fold:

1. My implementation of ERC-20 is more lightweight because it uses custom errors. Given the existing size of the pro contract, we care about contract size now.
2. It's not like OpenZeppelin is fixed in stone and it will never receive any further update. In fact, I have recently been contacted by the OZ team to share my feedback on their [EIP-6093](https://todoist.com/app/project/2284793622/task/6442563972), which aims to retroactively add custom errors to the popular token standards such as ERC-20. Therefore, even if we are to import OZ today with the goal of reducing the number of dependencies would-be Sablier integrators have to import, we will bump into the same type issue when OZ releases a new major version and everyone moves to that.

The solution for the type issue is the same as for the [PRBMath types](https://github.com/sablierhq/v2-core/issues/202) (thanks again @razgraf for the idea!) - to re-export the ERC-20 token interface in a dedicated types file, whatever the implementation ends up being.

**Comment by PaulRBerg** [comment]
Locking, we settled on using `IERC20` to avoid the potential confusion with `IERC721`.
